File: Stack And Queue/232.用栈实现队列.cpp
#include <stack>

using namespace std;
/*
 * @lc app=leetcode.cn id=232 lang=cpp
 *
 * [232] 用栈实现队列
 */

// @lc code=start
class MyQueue
{
public:
    stack<int> stIn;
    stack<int> stOut;
    /** Initialize your data structure here. */
    MyQueue()
    {
    }

    /** Push element x to the back of queue. */
    void push(int x)
    {
        // 输入保存元素
        stIn.push(x);
    }

    /** Removes the element from in front of queue and returns that element. */
    int pop()
    {
        // 反转输入栈，记录到输出栈
        if (stOut.empty())
        {
            while (!stIn.empty())
            {
                stOut.push(stIn.top());
                stIn.pop();
            }
        }

        int result = stOut.top();
        stOut.pop();

        return result;
    }

    /** Get the front element. */
    int peek()
    {
        int result = this->pop();
        stOut.push(result);

        return result;
    }

    /** Returns whether the queue is empty. */
    bool empty()
    {
        return stOut.empty() && stIn.empty();
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */
// @lc code=end

File: Stack And Queue/232.用栈实现队列.ts
/*
 * @lc app=leetcode.cn id=232 lang=typescript
 *
 * [232] 用栈实现队列
 */

// @lc code=start
class MyQueue {
    arr: any

    constructor() {
        this.arr = []
    }
    // 将元素 x 推到队列的末尾
    push(x: number): void {
        // 放入数组的尾部
        this.arr.push(x)
    }
    // 从队列的开头移除并返回元素
    pop(): number {
        // 移除数组头部元素，也就是先放进来的元素
        return this.arr.shift()
    }
    // 返回队列开头的元素
    peek(): number {
        if (this.empty()) {
            return null
        }

        // 移除数组头部元素，也就是先放进来的元素
        return this.arr[0]
    }
    //如果队列为空，返回 true ；否则，返回 false
    empty(): boolean {
        return this.arr.length === 0
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * var obj = new MyQueue()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.peek()
 * var param_4 = obj.empty()
 */
// @lc code=end

